nncf/utils.py

```diff
@@ -388,3 +388,47 @@ def compute_FLOPs_hook(module, input_, output, dict_to_save, ctx: 'TracingContex
 def add_domain(name_operator: str) -> str:
     from nncf.compression_method_api import DOMAIN_CUSTOM_OPS_NAME
     return DOMAIN_CUSTOM_OPS_NAME + "::" + name_operator
+
+
+def default_distributed_wrapper(model: nn.Module, execution_parameters: 'ExecutionParameters'):
+    """
+    Wrapping model for distributed training with DataParallel or DistributedDataParallel depending on execution mode
+    chosen by user.
+    :param execution_parameters: execution parameters
+    :param model: model to wrap  in accordance with execution mode chosen by user
+    :return: wrapped model
+    """
+    if not execution_parameters or execution_parameters.cpu_only:
+        # If execution params is not set or in cpu_only mode model can't be optimized by parallelization
+        return model
+
+    current_gpu = execution_parameters.current_gpu
+    if not is_dist_avail_and_initialized():
+        if current_gpu is not None:
+            # ExecutionMode.SINGLE_GPU
+            torch.cuda.set_device(current_gpu)
+        else:
+            # ExecutionMode.GPU_DATAPARALLEL
+            model = torch.nn.DataParallel(model)
+    else:
+        if current_gpu is None:
+            # ExecutionMode.DISTRIBUTED
+            model = torch.nn.parallel.DistributedDataParallel(model)
+        else:
+            # ExecutionMode.MULTIPROCESSING_DISTRIBUTED
+            torch.cuda.set_device(current_gpu)
+            model = torch.nn.parallel.distributed.DistributedDataParallel(model, device_ids=[current_gpu])
+    return model
+
+
+def default_distributed_unwrapper(model: nn.Module):
+    """
+    Unwrapping model prepared from distributed training in Pytorch
+    (and wrapped with Dataparallel or DistributedDataParallel).
+    :param model: model to unwrap.
+    :return: model without parallelization
+    """
+    if isinstance(model, (torch.nn.parallel.DataParallel, torch.nn.parallel.DistributedDataParallel)):
+        return model.module
+    else:
+        return model
```

Comment:
Is returning the underlying model enough? Will the future wrappings of the same model into DDP work as expected? I suppose that during each wrapping into DDP a certain context is created inside PyTorch that registers parameters; need to make sure that multiple wraps-unwraps made in this manner do not break the training (e.g. the parameters are not being registered or gradient-update twice etc.). This is something that should be determined from DDP docs and via testing.

Reply:
Tested it locally and everything works correctly.
Since such test isn't easy to create in our test infrastructure created task on this: 59185